### PR TITLE
Implementation of Smith normal form method.

### PR DESCRIPTION
**Pull request by AxelsFirst**
Added class of matrices, so that we can calculate Smith normal form of a matrix and it's rank.
Issue: #39

### CODE REVIEW

File: tda/matrix.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        """
+
+        Parameters:
+        -----------
+        entries: list
+            Two dimensional list of entries.
+
+        Notes:
+        ------
+        Entries have to be integers.
```
Comment (ghost) [inline tda/matrix.py]: Add a condition to check if the input is a list of non-empty lists of integers. An exception should be raised if something other than lists of integers is passed.

File: tda/matrix.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+        Output:
+        -------
+        row: list
+            A row.
+
+        """
+
+        return self.entries[row_index]
+
+    def get_col(self, col_index):
```
Comment (ghost) [inline tda/matrix.py]: Verify that `0 <= col_index < self.num_cols`.

File: tda/matrix.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+        ------
+        Entries have to be integers.
+
+        """
+
+        self.entries = entries
+        self.num_rows = len(entries)
+        self.num_cols = len(entries[0])
+
+    def get_row(self, row_index):
```
Comment (ghost) [inline tda/matrix.py]: Verify that `0 <= row_index < self.num_rows`.

File: tda/matrix.py [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+            A column.
+
+        """
+
+        col = []
+        for row in self.entries:
+            col.append(row[col_index])
+        return col
+
+    def change_row(self, row_index, values):
```
Comment (ghost) [inline tda/matrix.py]: Verify that `0 <= row_index < self.num_rows`.
Verify that `len(values) == len(self.entries[0])`.

File: tda/matrix.py [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+        -------
+        matrix: list
+            A matrix with changed rows.
+
+        """
+
+        self.entries[row_index] = values
+        return self.entries
+
+    def change_col(self, col_index, values):
```
Comment (ghost) [inline tda/matrix.py]: Verify that `0 <= col_index < self.num_cols`.
Verify that `len(values) == len(self.entries)`.

File: tda/matrix.py [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+        entries: list
+            A entries of a matrix with changed values.
+
+        """
+
+        for row_index in range(self.num_rows):
+            self.entries[row_index][col_index] = values[col_index]
+        return self.entries
+
+    def add_rows(self, index_row_to_change, index_row_to_add):
```
Comment (ghost) [inline tda/matrix.py]: Check correctness of `index_row_to_change` and `index_row_to_add`.

File: tda/matrix.py [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+        row_to_change = self.get_row(index_row_to_change)
+        row_to_add = self.get_row(index_row_to_add)
+        summed_row = []
+        for i in range(self.num_cols):
+            summed_entries = (row_to_change[i] + row_to_add[i]) % 2
+            summed_row.append(summed_entries)
+        self.change_row(index_row_to_change, summed_row)
+
+    def add_cols(self, index_col_to_change, index_col_to_add):
```
Comment (ghost) [inline tda/matrix.py]: Check correctness of `index_col_to_change` and `index_col_to_add`.

File: tda/matrix.py [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+        It will move a row, if non zero entry is on the same column as the
+        desired place.
+
+        It will move a columns, if non zero entry is right of
+        the desired place.
+
+        """
+
+        zero_row = True
+        for i in range(self.num_cols):
```
Comment (ghost) [inline tda/matrix.py]: Use `any(val != 0 for val in self.entries[entry_index])`.

File: tda/matrix.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+            return False
+
+        if self.entries[entry_index][entry_index] != 0:
+            return True
+
+        for row_index in chain(range(entry_index),
+                               range(entry_index+1, self.num_rows)):
+
+            if self.entries[row_index][entry_index] != 0:
+                copied_row = copy(self.get_row(row_index))
```
Comment (ghost) [inline tda/matrix.py]: Can we implement this without making copies? The cost of copying a list is O(n).

File: tda/matrix.py [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+
+                self.change_row(entry_index, copied_row)
+                self.change_row(row_index, copied_entry_row)
+
+                return True
+
+        for col_index in chain(range(entry_index),
+                               range(entry_index+1, self.num_cols)):
+            if self.entries[entry_index][col_index] != 0:
+                copied_col = copy(self.get_col(col_index))
```
Comment (ghost) [inline tda/matrix.py]: Can we implement this without making copies? The cost of copying a list is O(n).

File: tda/matrix.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+
+        """
+
+        snf = deepcopy(self)
+        max_dim = min(snf.num_rows, snf.num_cols)
+
+        for dim in range(max_dim):
+            value_check = snf.non_zero(dim)
+
+            if value_check is True:
```
Comment (ghost) [inline tda/matrix.py]: `if value_check:`

File: tda/matrix.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+            Smith normal form of a matrix.
+
+        Notes:
+        ------
+        The method uses modulo 2 addition.
+
+        The method is not consistent with theory, but it is good enough
+        for getting rank of a matrix.
+
+        """
```
Comment (ghost) [inline tda/matrix.py:338]: Maybe we should add examples?

File: tda/matrix.py [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+        ------
+        Intended to use only in smith_normal_form method due to algorithm
+        optimizations.
+
+        It will move a row, if non zero entry is on the same column as the
+        desired place.
+
+        It will move a columns, if non zero entry is right of
+        the desired place.
+
```
Comment (ghost) [inline tda/matrix.py:281]: Maybe we should add examples?

File: tda/matrix.py [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+        Output:
+        -------
+        snf_rank: int
+            Rank of Smith normal form of a matrix.
+
+        Notes:
+        ------
+        The method uses modulo 2 addition, with exception of calculating rank
+        itself.
+
```
Comment (ghost) [inline tda/matrix.py:367]: Maybe we should add examples?

File: tda/matrix.py [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+        Output:
+        -------
+        snf: matrix
+            Smith normal form of a matrix.
+
+        Notes:
+        ------
+        The method uses modulo 2 addition.
+
+        The method is not consistent with theory, but it is good enough
```
Comment (ghost) [inline tda/matrix.py]: Explain why it works if it's not consistent with theory.

**Review by ghost — CHANGES_REQUESTED** [review]
Minor comments pertaining mainly to checking inputs parameters, examples and documentation.

**Comment by AxelsFirst** [comment]
Requested changes have been implemented.

**Pull request merged**